packages/typescript-estree/src/useProgramFromProjectService.ts

```diff
@@ -0,0 +1,53 @@
+import path from 'path';
+import type * as ts from 'typescript';
+import type { server } from 'typescript/lib/tsserverlibrary';
+
+import { createProjectProgram } from './create-program/createProjectProgram';
+import {
+  type ASTAndDefiniteProgram,
+  getCanonicalFileName,
+} from './create-program/shared';
+import type { MutableParseSettings } from './parseSettings';
+
+export function useProgramFromProjectService(
+  projectService: server.ProjectService,
+  parseSettings: Readonly<MutableParseSettings>,
+): ASTAndDefiniteProgram | undefined {
+  const opened = projectService.openClientFile(
+    absolutify(parseSettings.filePath),
+    parseSettings.codeFullText,
+    /* scriptKind */ undefined,
+    parseSettings.tsconfigRootDir,
+  );
+  if (!opened.configFileName) {
+    return undefined;
+  }
+
+  const scriptInfo = projectService.getScriptInfo(parseSettings.filePath);
+  const program = projectService
+    .getDefaultProjectForFile(scriptInfo!.fileName, true)!
+    .getLanguageService(/*ensureSynchronized*/ true)
+    .getProgram();
+
+  if (!program) {
+    return undefined;
+  }
+
+  const configFilePath = program.getCompilerOptions().configFilePath as string;
+  if (
+    !parseSettings.projects
+      .map(absolutify)
+      .map(getCanonicalFileName)
+      .includes(getCanonicalFileName(configFilePath))
+  ) {
+    throw new Error(`Cannot read file '${parseSettings.projects[0]}'`);
+  }
```

Comment:
Is this bit needed? If I'm understanding this correctly, this still forces users to specify tsconfigs?

Reply:
We have a trio of tests that fail without it:
* `parseAndGenerateServices › projectFolderIgnoreList › ignores a folder when given a string glob`
* `semanticInfo › non-existent project file`
* `semanticInfo › fail to read project file`

...but I think you're right, with the project service is managing projects I don't see a need to preserve that behavior. I'll exclude them as well.